File: src/lib/api/pharmacy.ts
import apiClient, { ensureData } from '../api-client';

interface ApiResponse<T> {
  success: boolean;
  message?: string;
  data?: T;
}

// ========== 用藥錯誤回報 ==========

export interface ErrorReport {
  id: string;
  errorType: string;
  medicationName: string;
  description: string;
  patientId?: string | null;
  status: 'pending' | 'resolved';
  reporterId?: string | null;
  reporterName?: string | null;
  reporterRole?: string | null;
  actionTaken?: string | null;
  reviewedBy?: string | null;
  resolution?: string | null;
  timestamp: string;
  severity: 'low' | 'moderate' | 'high';
}

export interface ErrorReportsResponse {
  reports: ErrorReport[];
  total: number;
  stats?: {
    total: number;
    pending: number;
    resolved: number;
    byType: Record<string, number>;
    bySeverity: Record<string, number>;
  };
}

export interface ErrorReportsParams {
  page?: number;
  limit?: number;
  status?: string;
  type?: string;
}

export async function getErrorReports(params?: ErrorReportsParams): Promise<ErrorReportsResponse> {
  const response = await apiClient.get<ApiResponse<ErrorReportsResponse>>(
    '/pharmacy/error-reports',
    { params }
  );
  return ensureData(response.data, 'API contract');
}

export async function getErrorReportById(reportId: string): Promise<ErrorReport> {
  const response = await apiClient.get<ApiResponse<ErrorReport>>(
    `/pharmacy/error-reports/${reportId}`
  );
  return ensureData(response.data, 'API contract');
}

export interface CreateErrorReportData {
  errorType: string;
  medicationName: string;
  description: string;
  patientId?: string;
  actionTaken?: string;
  severity?: string;
}

export async function createErrorReport(
  data: CreateErrorReportData
): Promise<ErrorReport> {
  const response = await apiClient.post<ApiResponse<ErrorReport>>(
    '/pharmacy/error-reports',
    data
  );
  return ensureData(response.data, 'API contract');
}

export interface UpdateErrorReportData {
  status?: string;
  resolution?: string;
}

export async function updateErrorReport(
  reportId: string,
  data: UpdateErrorReportData
): Promise<ErrorReport> {
  const response = await apiClient.patch<ApiResponse<ErrorReport>>(
    `/pharmacy/error-reports/${reportId}`,
    data
  );
  return ensureData(response.data, 'API contract');
}

// ========== 相容性常用組合（持久化） ==========

export interface CompatibilityFavoritePair {
  id: string;
  drugA: string;
  drugB: string;
  solution: string;
  createdAt: string;
}

export interface CompatibilityFavoritesResponse {
  favorites: CompatibilityFavoritePair[];
  total: number;
}

export async function getCompatibilityFavorites(): Promise<CompatibilityFavoritesResponse> {
  const response = await apiClient.get<ApiResponse<CompatibilityFavoritesResponse>>(
    '/pharmacy/compatibility-favorites'
  );
  return ensureData(response.data, 'API contract');
}

export async function createCompatibilityFavorite(data: {
  drugA: string;
  drugB: string;
  solution?: string;
}): Promise<CompatibilityFavoritePair> {
  const response = await apiClient.post<ApiResponse<CompatibilityFavoritePair>>(
    '/pharmacy/compatibility-favorites',
    { drugA: data.drugA, drugB: data.drugB, solution: data.solution || 'none' }
  );
  return ensureData(response.data, 'API contract');
}

export async function deleteCompatibilityFavorite(favoriteId: string): Promise<void> {
  await apiClient.delete(`/pharmacy/compatibility-favorites/${favoriteId}`);
}

// ========== 交互作用與相容性查詢 ==========

export interface DrugInteractionSearchItem {
  id: string;
  drug1: string;
  drug2: string;
  severity: string;
  mechanism: string;
  clinicalEffect: string;
  management: string;
  references: string;
  riskRating?: string;
  riskRatingDescription?: string;
  severityLabel?: string;
  reliabilityRating?: string;
  routeDependency?: string;
  discussion?: string;
  footnotes?: string;
  dependencies?: string[];
  dependencyTypes?: string[];
  interactingMembers?: Array<{
    group_name: string;
    members: string[];
    exceptions: string[];
    exceptions_note: string;
  }>;
  pubmedIds?: string[];
}

export interface DrugInteractionSearchResponse {
  interactions: DrugInteractionSearchItem[];
  total: number;
}

export async function getDrugInteractions(params: {
  drugA: string;
  drugB?: string;
}): Promise<DrugInteractionSearchResponse> {
  const response = await apiClient.get<ApiResponse<DrugInteractionSearchResponse>>(
    '/pharmacy/drug-interactions',
    { params }
  );
  return ensureData(response.data, 'API contract');
}

export interface IVCompatibilitySearchItem {
  id: string;
  drug1: string;
  drug2: string;
  solution: string;
  compatible: boolean;
  timeStability?: string;
  notes?: string;
  references?: string;
}

export interface IVCompatibilitySearchResponse {
  compatibilities: IVCompatibilitySearchItem[];
  total: number;
}

export async function getIVCompatibility(params: {
  drugA: string;
  drugB: string;
  solution?: string;
}, options?: { suppressErrorToast?: boolean }): Promise<IVCompatibilitySearchResponse> {
  const response = await apiClient.get<ApiResponse<IVCompatibilitySearchResponse>>(
    '/pharmacy/iv-compatibility',
    { params, suppressErrorToast: options?.suppressErrorToast },
  );
  return ensureData(response.data, 'API contract');
}

// ========== 用藥建議統計 ==========

export interface AdviceStatistics {
  totalReports: number;
  resolvedRate: number;
  severityCounts: {
    low: number;
    moderate: number;
    high: number;
  };
}

/** @deprecated Misleading name — this endpoint queries ErrorReport (用藥錯誤通報), not PharmacyAdvice. */
export async function getAdviceStatistics(): Promise<AdviceStatistics> {
  const response = await apiClient.get<ApiResponse<AdviceStatistics>>(
    '/pharmacy/advice-statistics'
  );
  return ensureData(response.data, 'API contract');
}

// ========== 用藥建議記錄 ==========

export interface PharmacyAdviceRecord {
  id: string;
  patientId: string;
  patientName: string;
  bedNumber: string;
  adviceCode: string;
  adviceLabel: string;
  category: string;
  content: string;
  pharmacistName: string;
  timestamp: string;
  linkedMedications?: string[];
  accepted?: boolean | null;
}

export interface AdviceRecordsResponse {
  records: PharmacyAdviceRecord[];
  total: number;
}

export interface AdviceRecordsParams {
  month?: string;
  category?: string;
  page?: number;
  limit?: number;
}

export async function getAdviceRecords(params?: AdviceRecordsParams): Promise<AdviceRecordsResponse> {
  const response = await apiClient.get<ApiResponse<AdviceRecordsResponse>>(
    '/pharmacy/advice-records',
    { params }
  );
  return ensureData(response.data, 'API contract');
}

export interface CreateAdviceRecordData {
  patientId: string;
  adviceCode: string;
  adviceLabel: string;
  category: string;
  content: string;
  linkedMedications?: string[];
  accepted?: boolean;
}

export async function createAdviceRecord(data: CreateAdviceRecordData): Promise<PharmacyAdviceRecord> {
  const response = await apiClient.post<ApiResponse<PharmacyAdviceRecord>>(
    '/pharmacy/advice-records',
    data
  );
  return ensureData(response.data, 'API contract');
}

// ========== 用藥建議記錄統計（聚合） ==========

export interface AdviceRecordStats {
  total: number;
  byCategory: Array<{ category: string; count: number }>;
  byCode: Array<{ code: string; label: string; category: string; count: number }>;
  byPharmacist: Array<{ pharmacistName: string; count: number }>;
  byAcceptance?: { accepted: number; rejected: number; pending: number };
}

export async function getAdviceRecordStats(params?: { month?: string }): Promise<AdviceRecordStats> {
  const response = await apiClient.get<ApiResponse<AdviceRecordStats>>(
    '/pharmacy/advice-records/stats',
    { params }
  );
  return ensureData(response.data, 'API contract');
}

// ========== PAD 劑量計算 ==========

export interface PadDrugInfo {
  key: string;
  label: string;
  concentration: number;
  concentration_unit: string;
  dose_unit: string;
  dose_range: string;
  weight_basis: string;
  concentration_range?: [number, number];
}

export interface PadDrugsResponse {
  drugs: PadDrugInfo[];
}

export async function getPadDrugs(): Promise<PadDrugsResponse> {
  const response = await apiClient.get<ApiResponse<PadDrugsResponse>>(
    '/pharmacy/pad-drugs'
  );
  return ensureData(response.data, 'API contract');
}

export interface PadCalculateRequest {
  drug: string;
  weight_kg: number;
  target_dose_per_kg_hr: number;
  concentration: number;
  sex?: string;
  height_cm?: number;
}

export interface PadCalculateResult {
  drug: string;
  BMI?: number;
  IBW_kg?: number;
  AdjBW_kg?: number;
  pct_IBW?: number;
  is_obese?: boolean;
  weight_basis: string;
  dosing_weight_kg: number;
  dose_per_hr: number;
  rate_ml_hr: number;
  concentration: string;
  note?: string;
  steps: string[];
}

export async function padCalculate(data: PadCalculateRequest, options?: { suppressErrorToast?: boolean }): Promise<PadCalculateResult> {
  const response = await apiClient.post<ApiResponse<PadCalculateResult>>(
    '/pharmacy/pad-calculate',
    data,
    { suppressErrorToast: options?.suppressErrorToast },
  );
  return ensureData(response.data, 'API contract');
}

// 導出所有 API 函數
export const pharmacyApi = {
  getErrorReports,
  getErrorReportById,
  createErrorReport,
  updateErrorReport,
  getCompatibilityFavorites,
  createCompatibilityFavorite,
  deleteCompatibilityFavorite,
  getDrugInteractions,
  getIVCompatibility,
  getAdviceStatistics,
  getAdviceRecords,
  createAdviceRecord,
  getAdviceRecordStats,
  respondToAdvice,
};

// ========== 藥事建議回覆 ==========

export async function respondToAdvice(
  adviceRecordId: string,
  data: { accepted: boolean }
): Promise<PharmacyAdviceRecord> {
  const response = await apiClient.patch<ApiResponse<PharmacyAdviceRecord>>(
    `/pharmacy/advice-records/${adviceRecordId}/response`,
    data
  );
  return ensureData(response.data, 'API contract');
}

File: backend/app/schemas/pharmacy.py
from pydantic import BaseModel, Field


class CompatibilityFavoriteCreate(BaseModel):
    drugA: str = Field(..., min_length=1, max_length=200)
    drugB: str = Field(..., min_length=1, max_length=200)
    solution: str = Field("none", max_length=20)



# ── B2 batch 4 response schemas ──────────────────────────────────────────────

from datetime import datetime as _datetime
from typing import Any as _Any, Optional as _Optional

from pydantic import field_validator as _field_validator

from app.schemas.base import CamelModel as _CamelModel


class PharmacyAdviceResponse(_CamelModel):
    id: str
    patient_id: _Optional[str] = None
    patient_name: _Optional[str] = None
    bed_number: _Optional[str] = None
    advice_code: _Optional[str] = None
    advice_label: _Optional[str] = None
    category: _Optional[str] = None
    content: _Optional[str] = None
    pharmacist_name: _Optional[str] = None
    timestamp: _Optional[_datetime] = None
    linked_medications: list = []
    accepted: _Optional[bool] = None
    responded_by_id: _Optional[str] = None
    responded_by_name: _Optional[str] = None
    responded_at: _Optional[_datetime] = None

    @_field_validator("linked_medications", mode="before")
    @classmethod
    def _none_to_list(cls, v):
        return v or []


class CompatibilityFavoriteResponse(_CamelModel):
    id: str
    drug_a: _Optional[str] = None
    drug_b: _Optional[str] = None
    solution: _Optional[str] = None
    created_at: _Optional[_datetime] = None


class ErrorReportResponse(_CamelModel):
    id: str
    patient_id: _Optional[str] = None
    reporter_id: _Optional[str] = None
    reporter_name: _Optional[str] = None
    reporter_role: _Optional[str] = None
    error_type: _Optional[str] = None
    severity: _Optional[str] = None
    medication_name: _Optional[str] = None
    description: _Optional[str] = None
    action_taken: _Optional[str] = None
    status: _Optional[str] = None
    reviewed_by: _Optional[dict] = None
    resolution: _Optional[str] = None
    timestamp: _Optional[_datetime] = None


class IVCompatibilityRowResponse(_CamelModel):
    id: _Any = None
    drug1: _Optional[str] = None
    drug2: _Optional[str] = None
    solution: _Optional[str] = None
    compatible: _Optional[bool] = None
    time_stability: _Optional[str] = None
    notes: _Optional[str] = None
    references: _Optional[str] = None


class DrugInteractionRowResponse(_CamelModel):
    id: _Any = None
    drug1: _Optional[str] = None
    drug2: _Optional[str] = None
    severity: _Optional[str] = None
    mechanism: _Optional[str] = None
    clinical_effect: _Optional[str] = None
    management: _Optional[str] = None
    references: _Optional[str] = None
    risk_rating: _Optional[str] = None
    risk_rating_description: _Optional[str] = None
    severity_label: _Optional[str] = None
    reliability_rating: _Optional[str] = None
    route_dependency: _Optional[str] = None
    discussion: _Optional[str] = None
    footnotes: _Optional[str] = None
    dependencies: list = []
    dependency_types: list = []
    interacting_members: list = []
    pubmed_ids: list = []

    @_field_validator(
        "dependencies", "dependency_types", "interacting_members", "pubmed_ids",
        mode="before",
    )
    @classmethod
    def _json_field_to_list(cls, v):
        # 鏡射 _parse_json_field:str(JSON) 或 list 或 None → list
        if v is None:
            return []
        if isinstance(v, list):
            return v
        if isinstance(v, str):
            import json
            try:
                parsed = json.loads(v)
                return parsed if isinstance(parsed, list) else []
            except Exception:
                return []
        return []


class PharmacySoapRecordResponse(_CamelModel):
    """patientName 由 wrapper 依參數附加。"""
    id: str
    patient_id: _Optional[str] = None
    bed_number: _Optional[str] = None
    pharmacist_id: _Optional[str] = None
    pharmacist_name: _Optional[str] = None
    subjective: str = ""
    objective: str = ""
    assessment: str = ""
    plan: str = ""
    polished_content: str = ""
    created_at: _Optional[_datetime] = None
    updated_at: _Optional[_datetime] = None

    @_field_validator(
        "subjective", "objective", "assessment", "plan", "polished_content",
        mode="before",
    )
    @classmethod
    def _none_to_empty(cls, v):
        return v or ""
